Add permalink history for taxon on friendly-id

shahmayur001 opened the pull request:
### Description

This PR introduces `FriendlyId` to the `Taxon` model to store the history of permalinks (slugs) and provides a more flexible approach to handling slugs for taxons.

#### Changes:
- **Taxon Model**:
  - Added `FriendlyId` to the `Taxon` model, enabling it to generate and store slug history.
  - The `before_save` callback is used to set the permalink instead of `before_create` and `before_update`, allowing for better slug management.
  - Introduced a custom method for `should_generate_new_friendly_id?` to control when a new friendly ID should be generated, specifically when the permalink changes.
  - Added a custom `normalize_friendly_id` method to adjust how slugs are generated, especially for nested taxons, ensuring proper formatting of the slug.

- **Test**:
  - Added a test to ensure that the history of permalinks is correctly stored when a taxon's permalink is updated. The test verifies that both the initial and updated slugs are retained in the history.

### Motivation
This update enhances the management of taxon slugs by allowing historical tracking of permalinks. This is particularly useful for SEO purposes and for maintaining stable URLs even if taxon names change over time.

elia commented on core/spec/models/spree/taxon_spec.rb:
```diff
@@ -269,4 +269,25 @@
+      expect(taxon.slugs.last.slug).to eq(initial_permalink)
+
+      taxon.update(permalink: "rails for ruby")
+      updated_permalink = taxon.permalink
+
+      expect(taxon.permalink).to eq(updated_permalink)
```
This is a bit of a tautology, I suggest putting the string itself in the expectation without relying on dynamic data, that makes for a much better and robust spec.

shahmayur001 replied:
@elia That make sense, Thank you!

shahmayur001 replied:
@elia Resolved!

fthobe commented:
@shahmayur001 as a consequence of this change, do we need to write logic to redirect {id} and {historyslug} as we did on products?

fthobe commented:
@tvdeyen as bespoken history for Taxon slugs as well

tvdeyen reviewed (commented):
I explained it many times before and still I need to comment on **every single PR of yours**. 

Please use your (often very good PR descriptions) as commit description(s). It is not even more work for you, because GitHub actually uses the git commit message to create the PR description in single commit PRs like this.

So, please **always** use the commit description for explaining **why a change has been** made. Even for multi commit PRs. The PR description is only helpful for users of the GH UI, while **commit messages are used every where else** (in your code editor, in third party git clients, the git log in your terminal, during rebase, etc.). It should always be **the single source of truth and not the GitHub GUI**). This has been discussed in the developer community **all over the globe for decades** now and should have been inherited by every single developer finally. I am a bit frustrated that I have to repeat this every, single time.

So, please - again - rewrite your commit message so that it includes the content of your - actually very good - PR description.

I hope that I do not have to write **the same comment on your PRs** over and over again. 

Thanks.

shahmayur001 commented:
> I explained it many times before and still I need to comment on **every single PR of yours**.
> 
> Please use your (often very good PR descriptions) as commit description(s). It is not even more work for you, because GitHub actually uses the git commit message to create the PR description in single commit PRs like this.
> 
> So, please **always** use the commit description for explaining **why a change has been** made. Even for multi commit PRs. The PR description is only helpful for users of the GH UI, while **commit messages are used every where else** (in your code editor, in third party git clients, the git log in your terminal, during rebase, etc.). It should always be **the single source of truth and not the GitHub GUI**). This has been discussed in the developer community **all over the globe for decades** now and should have been inherited by every single developer finally. I am a bit frustrated that I have to repeat this every, single time.
> 
> So, please - again - rewrite your commit message so that it includes the content of your - actually very good - PR description.
> 
> I hope that I do not have to write **the same comment on your PRs** over and over again.
> 
> Thanks.

Thank you for the feedback. I was planning to update the commit message after addressing the comments but I completely agree with your point. Moving forward, I will ensure that commit includes the description. I appreciate your patience.

tvdeyen commented:
> Thank you for the feedback. I was planning to update the commit message after addressing the comments but I completely agree with your point. Moving forward, I will ensure that commit includes the description. I appreciate your patience.

Thanks for your understanding. I actually is very helpful to have all the information beforehand in order to be able to review the PR in the first place.

fthobe commented:
@tvdeyen thank you

fthobe commented:
@tvdeyen there was a question on top of the PR, could you please answer to that so that we know what to do there. I understand that this is a breaking change, but I feel like it should be tackled sooner or later as current naming conventions between products and taxons regarding URLs are confusing.

tvdeyen commented:
We can do that, but is it worth the effort?

Because this is would we would have to do

1. Introduce the new column
2. Deprecate the old column
3. Write data into both columns
4. Stop writing the data into the old column
5. Drop old column

All that over the period of at least one major version. I am not sure if this is really worth it. What is the reasoning behind this? Just because or is there a technical issue that this solves?

fthobe commented:
> We can do that, but is it worth the effort?
> 
> Because this is would we would have to do
> 
> 1. Introduce the new column
> 2. Deprecate the old column
> 3. Write data into both columns
> 4. Stop writing the data into the old column
> 5. Drop old column
> 
> All that over the period of at least one major version. I am not sure if this is really worth it. What is the reasoning behind this? Just because or is there a technical issue that this solves?

In this case absolutely only bringing naming in line between resources. I wouldn't die on that hill. Taxon rendered is taxon rendered no matter the name.

tvdeyen commented:
Same here. I don't think it's worth it. 

Thanks for elaborating
